lib/vdsm/virt/livemerge.py

```diff
@@ -432,21 +438,53 @@ def _base_needs_extend(self, drive, job, base_info):
             log.debug("Base volume does not support extending "
                       "job=%s drive=%s volume=%s",
                       job.id, job.drive, job.base)
-            return False
+            return False, None
 
+        current_size = int(base_info["apparentsize"])
         max_size = drive.getMaxVolumeSize(int(base_info["capacity"]))
 
         # Size can be bigger than maximum value due to rounding to LVM extent
         # size (128 MiB).
-        if int(base_info["apparentsize"]) >= max_size:
+        if current_size >= max_size:
             log.debug("Base volume is already extended to maximum size "
                       "job=%s drive=%s volume=%s size=%s",
                       job.id, job.drive, job.base, base_info["apparentsize"])
-            return False
+            return False, None
+
+        # Measure the required base size for this merge, based on the current
+        # allocation in top. This measurement is racy, since the guest may
+        # write data to top after the measure.
+        measure = self._vm.measure(
+            drive.domainID,
+            drive.imageID,
+            job.top,
+            base_info["format"],
+            baseID=job.base)
+
+        # Consider the required size and possible bitmaps. Since some of the
+        # bitmaps already exist in base, this allocates more space than needed,
+        # but bitmaps are small so this is good enough.
+        required_size = measure["required"] + measure.get("bitmaps", 0)
+
+        chunk_size = config.getint("irs", "volume_utilization_chunk_mb") * MiB
+
+        if job.active_commit:
+            # For active layer merge, add one chunk of free space to avoid
+            # pausing during extend, or right after extend completes. This is
+            # a heuristic, we cannot prevent pausing during merge without
+            # monitoring the volume.
+            if required_size + chunk_size > current_size:
+                return True, required_size
```

Comment:
Shouldn't we return here `True, required_size + chunk_size`? If for some reason not and `required_size` is enough, if would IMHO deserve small explanation in the comment above.